### ISSUE TITLE
RFC: refactoring the internal

### ISSUE THREAD

**Issue by shinima** [redux-saga/redux-saga#1483]
tldr: I am planning to make some changes to refactor the redux-saga internal which includes removing `runSaga#options.dispatch` / `runSaga#options.getState` and seperating effect-runners into categories, how do you think about it?

I am excited about saga running environments introduced in #1479. `env` brings more opportunities to simplify `proc()` and make it easier to extend redux-saga. `proc()` is over complex now (the function definition span 500+ lines), I am planning to make use of `env` and move complexity from `proc()` to `runSaga()`.

#1481 is an attempt to simplify the implementation of `digestEffect`. In that PR, I extract the logic of composing middlewares to `env.finalizeRunEffect` (or `env.enhanceRunEffect`) in runSaga. For `proc()`, getting a `env.finalizeRunEffect` instead of `env.middleware` means reducing the knowledge requirement about the outter world.

As [the doc says](https://redux-saga.js.org/docs/advanced/UsingRunSaga.html), `runSaga` could be used to connect external input/output, but currently `runSaga` requires `options.dispatch` and `options.getState`, which means our `runSaga/proc` is still coupled with redux since `dispatch` and `getState` are APIs from redux. Ideally, `runSaga` should not require any redux-related options. Removing `options.dispatch` and `options.getState` makes it easier to run sagas in non-redux environments.

In 2fb401d68afc988563daff608d210e9a8d295cf7, I remove `env.dispatch` by passing `runSaga` a special channel in *src/internal/middleware.js*. The special channel could be thought as a redux-store-connected stdChannel: every put effect to this channel will trigger `store.dispatch`, and every call of `store.dispatch` will put the action to this channel.

Removing `env.getState` is difficult now, because select effects depend on `env.getState`. Removing `env.getState` requires a mechanism that allows defining new effect types when `runSaga` is called. `env.runCustomEffect` seems like a good option now, we can use it internally first to remove `env.getState`. That means running logic of select effects is defined in *src/internal/middleware.js*, and select effects are only available in sagas started by `sagaMiddleware.run`. Later on, I'll try to find better alternatives to extend effect types, and perhaps we could allow extending in the user land.

We have tried move effect-runner out of `proc` in #1263, but it has lots of conflicts and can hardly merge now. I've implemented the separation in my [little-saga](https://github.com/shinima/little-saga). Here is an image showing the structure of little-saga:

![image](https://user-images.githubusercontent.com/5550931/41168748-7b5699ac-6b79-11e8-881b-a6a362241677.png)

The code in folder *src/core/* implements the core effect types, which inlcudes `promise / iterator / fork / spawn / join / cancel / cancelled / def`. Other effect types are extended by def-effect. *src/commonEffects* implements `all / race / apply / call / cps / getContext / setContext`, and `src/channelEffects` implementes channel-related effects including `put / take / flush / actionChannel`. In little-saga, select effect is defined when connecting to a redux store.

Once `env.runCustomEffect` can be use internally, I think we could seperate redux-saga's effect-runners into several categories like in little-saga. Redux-saga is much more complicated than little-saga, I'll write code carefully and try to avoid breaking changes (expect for `runSaga#options.getState` and `runSaga#options.dispatch`).

How do you think about these changes? If you agree my proposal, I am glad to bring these changes. I am open to any suggestion, so just tell me any idea/thought you have.

**Comment by Andarist** [comment]
> As the doc says, runSaga could be used to connect external input/output, but currently runSaga requires options.dispatch and options.getState, which means our runSaga/proc is still coupled with redux since dispatch and getState are APIs from redux. Ideally, runSaga should not require any redux-related options. Removing options.dispatch and options.getState makes it easier to run sagas in non-redux environments.

We can certainly discuss alternative API, but even though dispatch/getState names are obviously taken from redux you can pass any functions as those to connect external input/output. Don't quite think this is a blocker for anyone if they need this lower level API.

At the moment I don't particularly feel like removing internally used dispatch/getState is an important thing to do, I don't see it making things any easier. They are kinda a nicely defined way of integrating saga with anything - where `getState` is used for querying something and `dispatch` is acting like an "emit" function. 

What is more interesting to me is to see how exactly have you separated various effect categories in your little-saga. I will have to take a look into that in the following week.

Don't be discouraged by my words though, I'm happy to discuss any idea.

**Comment by shinima** [comment]
> Don't quite think this is a blocker for anyone if they need this lower level API.
Yes, you are right. I've made mistakes. I was thinking that `dispatch / getState` introduces much challenge for connecting external I/O ( emitter, Observable). But after a second thinking, it is fine that using `dispatch / getState` actually.

> What is more interesting to me is to see how exactly have you separated various effect categories in your little-saga. I will have to take a look into that in the following week.
I can make a trial PR in the next few days.
